src/worker.cc

```diff
@@ -1233,12 +1089,93 @@ worker_t::result_t to_response(const std::string& data,
+};
+
+service_worker_t::service_worker_t(const boost::property_tree::ptree& config)
+    : interrupt(nullptr), statsd_client(new statsd_client_t(config)) {
 }
 service_worker_t::~service_worker_t() {
 }
 void service_worker_t::set_interrupt(const std::function<void()>* interrupt_function) {
   interrupt = interrupt_function;
 }
+void service_worker_t::cleanup() {
+  // sends metrics to statsd server over udp
+  statsd_client->flush();
+}
+void service_worker_t::enqueue_statistics(Api& api) const {
+  // nothing to do without stats
+  if (!api.has_info() || api.info().statistics().empty())
+    return;
+
+  // these have been filled out as the request progressed through the system
+  for (const auto& stat : api.info().statistics()) {
+    float frequency = stat.has_frequency() ? stat.frequency() : 1.f;
+    switch (stat.type()) {
+      case count:
+        statsd_client->count(stat.key(), static_cast<int>(stat.value() + 0.5), frequency,
+                             statsd_client->tags);
+        break;
+      case gauge:
+        statsd_client->gauge(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                             statsd_client->tags);
+        break;
+      case timing:
+        statsd_client->timing(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                              statsd_client->tags);
+        break;
+      case set:
+        statsd_client->set(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                           statsd_client->tags);
+        break;
+    }
+  }
+
+  // before we are done with the request, if this was not an error we log it was ok
+  if (api.info().has_error() && api.info().error()) {
+    auto worker = typeid(*this) == typeid(loki::loki_worker_t)
+                      ? ".loki."
+                      : (typeid(*this) == typeid(thor::thor_worker_t) ? ".thor." : ".odin.");
+    const auto& action = Options_Action_Enum_Name(api.options().action());
+
+    statsd_client->count(action + worker + "ok", 1, 1.f, statsd_client->tags);
+  }
+}
+midgard::Finally<std::function<void()>> service_worker_t::measure_scope_time(Api& api) const {
+  // we copy the captures that could go out of scope
+  auto start = std::chrono::steady_clock::now();
+  return midgard::Finally<std::function<void()>>([this, &api, start]() {
+    auto elapsed = std::chrono::steady_clock::now() - start;
+    auto e = std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(elapsed).count();
+    auto worker = typeid(*this) == typeid(loki::loki_worker_t)
+                      ? ".loki."
+                      : (typeid(*this) == typeid(thor::thor_worker_t) ? ".thor." : ".odin.");
+    const auto& action = Options_Action_Enum_Name(api.options().action());
+
+    auto* stat = api.mutable_info()->mutable_statistics()->Add();
+    stat->set_key(action + worker + "latency_ms");
+    stat->set_value(e);
+    stat->set_type(timing);
+  });
+}
```

Comment:
ported the `scoped_timer` stuff to use the `Finally` stuff. i made it a member function so it could use `this` to do type checkign to figure out what kind of worker it was, which precluded it from being called from within serialization (hence the change to odin_worker::narrate)